pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,386 @@
+		if *progressing {
+			progressingText = "waiting for CSR to be signed"
+			progressingStatus = operatorv1.ConditionTrue
+			progressingReason = "WaitingForCertificate"
+		}
+
+		_, _, statusErr := v1helpers.UpdateStatus(c.operatorClient,
+			v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "AuthenticatorCSRProgressing",
+				Status:  progressingStatus,
+				Reason:  progressingReason,
+				Message: progressingText,
+			}))
+
+		if statusErr != nil {
+			if err != nil {
+				klog.Error("failed to update operator status: %v", statusErr)
+				return err
+			}
+			return statusErr
+		}
+	}
+
+	return err
+}
+
+func (c *webhookAuthenticatorCertRequester) sync(ctx context.Context, syncCtx factory.SyncContext) (progressing *bool, err error) {
+	key, cert, keyRotated, err := c.getKeyCertFromSecret(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if keyRotated {
+		if err := c.resubmitCSR(ctx, key); err != nil {
+			return nil, err
+		}
+		return pbool(true), nil
+	}
+
+	if cert != nil {
+		// check that the certificate is not expiring
+		certValidityDuration := cert.NotAfter.Sub(cert.NotBefore)
+		if time.Now().Before(cert.NotBefore.Add(time.Duration(0.8 * float64(certValidityDuration)))) {
+			return pbool(false), nil
+		}
+
+		expiryMsg := fmt.Sprintf("the authenticator's certificate is almost expired (%s), will schedule a new CSR creation", cert.NotAfter.String())
+		syncCtx.Recorder().Eventf("ClientCertExpiringSoon", expiryMsg)
+		klog.V(5).Infof(expiryMsg)
+
+		// delete the cert/key secret and schedule a new sync
+		if err := c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{}); err != nil {
+			return nil, err
+		}
+		return nil, factory.SyntheticRequeueError
+	}
+
+	// either we don't have a cert yet or it's almost expired
+	csrObj, err := c.csrLister.Get(csrName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if csrObj != nil {
+		if len(csrObj.Status.Conditions) == 0 {
+			return pbool(true), nil
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			switch cond.Type {
+			case certapiv1.CertificateDenied, certapiv1.CertificateFailed:
+				if cond.Status == corev1.ConditionTrue {
+					syncCtx.Recorder().Eventf("CSRFailure", "previous CSR got denied or it otherwise failed: {%s - %s: %s}", cond.Type, cond.Reason, cond.Message)
+					klog.Warning("the previous CSR got denied or it otherwise failed: %v", cond)
+					if err := c.resubmitCSR(ctx, key); err != nil {
+						return nil, err
+					}
+					return pbool(true), nil
+				}
+			}
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			if cond.Type == certapiv1.CertificateApproved && cond.Status == corev1.ConditionTrue {
+				csrCert := csrObj.Status.Certificate
+				if len(csrCert) == 0 {
+					return pbool(true), nil
+				}
+				certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+				if err != nil {
+					return nil, err
+				}
+				certSecret.Data["tls.crt"] = csrCert
```

Comment:
how do you make secrets immutable with what we've got today?

Reply:
https://kubernetes.io/docs/concepts/configuration/secret/#secret-immutable

Reply:
Interesting, I was looking for "sealed secrets" but that brought something else. I did not know about this. Shouldn't we wait until it goes stable before we start using it, though?

Reply:
it went `GA` in 1.21 - https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1412-immutable-secrets-and-configmaps/kep.yaml#L31

Reply:
awesome, I'll try to rework the controller then

Reply:
just a suggestion, it is okay if you don't like it :)

Reply:
oh no, I like the suggestion. Although, as I started implementing it I noticed, that it may not bring that much value adding the immutable field in here. You still have no way of preventing someone to stopping the controller, replacing the secret, and starting it again, so you still need to perform all the checks.